File: WeatherSense/trunk/VantagePro2Driver/src/vantagepro2/ArchivePacket.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;

namespace VantagePro2 {
    /// <summary>
    /// Class that holds a single archive packet. It also formats the WeatherSense XML message.
    /// </summary>
    public class ArchivePacket {
        /// <summary>
        /// Time property
        /// </summary>
        /// <value>The time that this archive packet was recorded</value>
        private DateTime time;
        public DateTime Time {
            get { return time; }
        }

        private int windSampleCount;
        public int WindSampleCount {
            get { return windSampleCount; }
        }
        
        public static readonly int BYTES_PER_RECORD = 52;
        private int archivePeriod;
        private float rainInterval;
        private float windGust;
        private int windGustDirection;

        private byte[] buffer = new byte[BYTES_PER_RECORD];
        public byte[] Buffer {
            get { return buffer; }
        }

        public static readonly int NUM_EXTRA_TEMPERATURES = 3;
        public static readonly int NUM_LEAF_TEMPERATURES = 2;
        public static readonly int NUM_LEAF_WETNESSES = 2;
        public static readonly int NUM_SOIL_TEMPERATURES = 3;
        public static readonly int NUM_EXTRA_HUMIDITIES = 2;
        public static readonly int NUM_SOIL_MOISTURES = 4;
        private static readonly int UNKNOWN_TEMPERATURE = 32767;
        private static readonly int UNKNOWN_HIGH_TEMPERATURE = -32768;
        private static readonly int UNKNOWN_SOLAR_RADIATION = 32767;
        private static readonly int UNKNOWN_HUMIDITY = 255;
        private static readonly int UNKNOWN_WIND_DIRECTION = 255;
        private static readonly int UNKNOWN_WIND_SPEED = 255;
        private static readonly int UNKNOWN_PRESSURE = 0;
        private static readonly int UNKNOWN_UV_INDEX = 255;
        private static readonly int UNKNOWN_EXTRA_TEMPERATURE = 255;
        private static readonly int UNKNOWN_LEAF_WETNESS = 255;
        private static readonly int UNKNOWN_SOIL_MOISTURE = 255;
        private static readonly int UNKNOWN_ET = 0;
        private static readonly float TEMPERATURE_SCALE = 10.0F;
        private static readonly float BAROMETER_SCALE = 1000.0F;
        private static readonly float UV_INDEX_SCALE = 10.0F;
        private static readonly float ET_SCALE = 1000.0F;
        private static readonly int EXTRA_TEMPERATURE_OFFSET = 90;
        private static readonly float DEGREES_PER_SLICE = 22.5F;

        /// <summary>
        /// Constuctor.
        /// </summary>
        /// <param name="buffer">The buffer from which to extract the data for the archive packet</param>
        /// <param name="offset">The offset within the buffer that the packet starts</param>
        /// <param name="rainInterval">The amount of rain each tip of the rain bucket represents</param>
        /// <param name="archivePeriod">The archive period (in minutes)</param>
        /// <param name="windGust">The current wind gust value from the loop 2 packet</param>
        /// <param name="windGustDirection">The current wind gust direction from the loop 2 packet</param>
        public ArchivePacket(byte[] buffer, int offset, float rainInterval, int archivePeriod, float windGust, int windGustDirection) {
            this.archivePeriod = archivePeriod;
            this.rainInterval = rainInterval;
            this.windGust = windGust;
            this.windGustDirection = windGustDirection;
            for (int i = 0; i < BYTES_PER_RECORD; i++) {
                this.buffer[i] = buffer[offset + i];
            }

            windSampleCount = BitConverter.ToInt16(this.buffer, 18);
            time = extractDate();
        }

        private DateTime extractDate() {
            short date = BitConverter.ToInt16(buffer, 0);
            short time = BitConverter.ToInt16(buffer, 2);
            int year = ((date >> 9) & 0x3F) + 2000;
            int month = (date >> 5) & 0xF;
            int day = date & 0x1F;
            int hour = time / 100;
            int minute = time % 100;

            DateTime dateTime = new DateTime(year, month, day, hour, minute, 0, DateTimeKind.Local);
            return dateTime;
        }

        /// <summary>
        /// Format the WeatherSense historical record XML message.
        /// <para>The message will only include those values that are valid (not dashed)</para>
        /// </summary>
        /// <returns>The message</returns>
        public String formatMessage() {
            StringBuilder sb = new StringBuilder("<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>");
            sb.Append("<historicalRecord>");
            DateTime time = extractDate();
            sb.AppendFormat("<time>{0:O}</time>", time);
            sb.AppendFormat("<duration>{0}</duration>", archivePeriod * 60);

            short value16 = BitConverter.ToInt16(buffer, 4);
            if (value16 != UNKNOWN_TEMPERATURE)
                sb.AppendFormat("<avgOutdoorTemperature><value>{0}</value></avgOutdoorTemperature>", UnitConverter.toCelcius((float)value16 / TEMPERATURE_SCALE));

            value16 = BitConverter.ToInt16(buffer, 6);
            if (value16 != UNKNOWN_HIGH_TEMPERATURE)
                sb.AppendFormat("<highOutdoorTemperature><value>{0}</value></highOutdoorTemperature>", UnitConverter.toCelcius((float)value16 / TEMPERATURE_SCALE));

            value16 = BitConverter.ToInt16(buffer, 8);
            if (value16 != UNKNOWN_TEMPERATURE)
                sb.AppendFormat("<lowOutdoorTemperature><value>{0}</value></lowOutdoorTemperature>", UnitConverter.toCelcius((float)value16 / TEMPERATURE_SCALE));

            value16 = BitConverter.ToInt16(buffer, 10);
            sb.AppendFormat("<rainfall><value>{0}</value></rainfall>", UnitConverter.toMillimeter((float)value16 * rainInterval));

            value16 = BitConverter.ToInt16(buffer, 12);
            sb.AppendFormat("<highRainfallRate><value>{0}</value></highRainfallRate>", UnitConverter.toMillimeter((float)value16 * rainInterval));

            value16 = BitConverter.ToInt16(buffer, 14);
            if (value16 != UNKNOWN_PRESSURE)
                sb.AppendFormat("<baroPressure><value>{0}</value></baroPressure>", UnitConverter.toInchesHg((float)value16 / BAROMETER_SCALE));

            value16 = BitConverter.ToInt16(buffer, 16);
            if (value16 != UNKNOWN_SOLAR_RADIATION)
                sb.AppendFormat("<avgSolarRadiation><value>{0}</value></avgSolarRadiation>", value16);
          
            value16 = BitConverter.ToInt16(buffer, 20);
            if (value16 != UNKNOWN_TEMPERATURE)
                sb.AppendFormat("<indoorTemperature><value>{0}</value></indoorTemperature>", UnitConverter.toCelcius((float)value16 / TEMPERATURE_SCALE));

            byte value8 = buffer[22];
            if (value8 != UNKNOWN_HUMIDITY)
                sb.AppendFormat("<indoorHumidity><value>{0}</value></indoorHumidity>", value8);

            value8 = buffer[23];
            if (value8 != UNKNOWN_HUMIDITY)
                sb.AppendFormat("<outdoorHumidity><value>{0}</value></outdoorHumidity>", value8);

            int windSpeed = buffer[24];
            if (windSpeed == UNKNOWN_WIND_SPEED)
                windSpeed = 0;
            int windDirection = buffer[27];
            sb.AppendFormat("<avgWind><speed><value>{0}</value></speed>", UnitConverter.toMetersPerSecond(windSpeed));
            if (windDirection != UNKNOWN_WIND_DIRECTION)
                sb.AppendFormat("<direction><value>{0}</value></direction>", windDirection * DEGREES_PER_SLICE);
            sb.Append("</avgWind>");

            windSpeed = buffer[25];
            if (windSpeed == UNKNOWN_WIND_SPEED)
                windSpeed = 0;
            windDirection = buffer[26];
            sb.AppendFormat("<highWind><speed><value>{0}</value></speed>", UnitConverter.toMetersPerSecond(windSpeed));
            if (windDirection != UNKNOWN_WIND_DIRECTION)
                sb.AppendFormat("<direction><value>{0}</value></direction>", windDirection * DEGREES_PER_SLICE);
            sb.Append("</highWind>");

            
            //
            // Use the wind gust data from the LOOP 2 packet if this archive packet is less than 10 minutes old,
            // otherwise just use the high wind speed as the gust
            //
            DateTime now = DateTime.Now;
            TimeSpan span = now.Subtract(time);
            if (span.TotalMinutes < 10) {
                if (windGust != 0.0)
                    sb.AppendFormat("<windGust><speed><value>{0}</value></speed><direction><value>{1}</value></direction></windGust>", windGust, windGustDirection);
            }
            else {
                windSpeed = buffer[25];
                if (windSpeed == UNKNOWN_WIND_SPEED)
                    windSpeed = 0;
                windDirection = buffer[26];
                sb.AppendFormat("<windGust><speed><value>{0}</value></speed>", UnitConverter.toMetersPerSecond(windSpeed));
                if (windDirection != UNKNOWN_WIND_DIRECTION)
                    sb.AppendFormat("<direction><value>{0}</value></direction>", windDirection * DEGREES_PER_SLICE);
                sb.Append("</windGust>");
            }

            value8 = buffer[28];
            if (value8 != UNKNOWN_UV_INDEX)
                sb.AppendFormat("<avgUvIndex><value>{0}</value></avgUvIndex>", (float)value8 / UV_INDEX_SCALE);

            value8 = buffer[29];
            if (value8 != UNKNOWN_ET)
                sb.AppendFormat("<evapotranspiration><value>{0}</value></evapotranspiration>", UnitConverter.toMillimeter((float)value8 / ET_SCALE));

            value16 = BitConverter.ToInt16(buffer, 30);
            if (value16 != UNKNOWN_SOLAR_RADIATION)
                sb.AppendFormat("<highSolarRadiation><value>{0}</value></highSolarRadiation>", value16);

            value8 = buffer[32];
            if (value8 != UNKNOWN_UV_INDEX)
                sb.AppendFormat("<highUvIndex><value>{0}</value></highUvIndex>", (float)value8 / UV_INDEX_SCALE);

            /*
            for (int i = 0; i < NUM_LEAF_TEMPERATURES; i++) {
                value8 = buffer[34 + i];
                if (value8 != UNKNOWN_EXTRA_TEMPERATURE)
                    sb.AppendFormat("leaf_temperature[{0}]={1};", i, value8 - EXTRA_TEMPERATURE_OFFSET);
            }
            */


            sb.Append("<leafWetnessSensorEntries>");
            for (int i = 0; i < NUM_LEAF_WETNESSES; i++) {
                int leafWetness = buffer[36 + i];
                if (leafWetness != UNKNOWN_LEAF_WETNESS) {
                    sb.AppendFormat("<entry><key>{0}</key><value><sensorId>{1}</sensorId><sensorType>LEAF_WETNESS</sensorType>", 500 + i, 500 + i);
                    sb.Append("<measurement xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xsi:type=\"leafWetness\">");
                    sb.AppendFormat("<value>{0}</value></measurement></value></entry>", leafWetness);
                }
            }
            sb.Append("</leafWetnessSensorEntries>");

            /*
            for (int i = 0; i < NUM_SOIL_TEMPERATURES; i++) {
                value8 = buffer[38 + i];
                if (value8 != UNKNOWN_EXTRA_TEMPERATURE)
                    sb.AppendFormat("soil_temperature[{0}]={1};", i, value8 - EXTRA_TEMPERATURE_OFFSET);
            }
            */

            sb.Append("<humiditySensorEntries>");
            for (int i = 0; i < NUM_EXTRA_HUMIDITIES; i++) {
                int humidity = buffer[43 + i];
                if (humidity != UNKNOWN_HUMIDITY) {
                    sb.AppendFormat("<entry><key>{0}</key><value><sensorId>{1}</sensorId><sensorType>HYGROMETER</sensorType>", 200 + i, 200 + i);
                    sb.Append("<measurement xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xsi:type=\"humidity\">");
                    sb.AppendFormat("<value>{0}</value></measurement></value></entry>", humidity);
                }
            }
            sb.Append("</humiditySensorEntries>");

            sb.Append("<temperatureSensorEntries>");
            for (int i = 0; i < NUM_EXTRA_TEMPERATURES; i++) {
                int temperature = buffer[45 + i];
                if (temperature != UNKNOWN_EXTRA_TEMPERATURE) {
                    sb.AppendFormat("<entry><key>{0}</key><value><sensorId>{1}</sensorId><sensorType>THERMOMETER</sensorType>", 100 + i, 100 + i);
                    sb.Append("<measurement xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xsi:type=\"temperature\">");
                    sb.AppendFormat("<value>{0}</value></measurement></value></entry>", UnitConverter.toCelcius(temperature - EXTRA_TEMPERATURE_OFFSET));
                }
            }
            sb.Append("</temperatureSensorEntries>");

            sb.Append("<soilMoistureSensorEntries>");
            for (int i = 0; i < NUM_SOIL_MOISTURES; i++) {
                int soilMoisture = buffer[48 + i];
                if (soilMoisture != UNKNOWN_SOIL_MOISTURE) {
                    sb.AppendFormat("<entry><key>{0}</key><value><sensorId>{1}</sensorId><sensorType>SOIL_MOISTURE</sensorType>", 600 + i, 600 + i);
                    sb.Append("<measurement xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xsi:type=\"soilMoisture\">");
                    sb.AppendFormat("<value>{0}</value></measurement></value></entry>", soilMoisture);
                }
            }
            sb.Append("</soilMoistureSensorEntries>");
            sb.Append("</historicalRecord>");
            return sb.ToString();
        }

        public override String ToString() {
            return formatMessage();
        }
    }
}

File: VantagePro2Driver/VP2++/src/VP2/ArchivePacket.cpp
/* 
 * Copyright (C) 2022 Bruce Beisel
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
#include <time.h>
#include <string>
#include <ctime>
#include <sstream>
#include "VP2Logger.h"
#include "VP2Constants.h"
#include "VP2Decoder.h"
#include "BitConverter.h"
#include "UnitConverter.h"
#include "ArchivePacket.h"

using namespace std;

namespace vp2 {

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
ArchivePacket::ArchivePacket() : packetTime(0), windSampleCount(0), buffer(""), log(&VP2Logger::getLogger("ArchivePacket")) {
}


////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
ArchivePacket::ArchivePacket(const byte buffer[], int offset) : log(&VP2Logger::getLogger("ArchivePacket")) {
    updateArchiveData(buffer, offset);
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
ArchivePacket::~ArchivePacket() {
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
void
ArchivePacket::updateArchiveData(const byte buffer[], int offset) {
    //
    // Copy the packet from the passed in buffer to the buffer member
    //
    for (int i = 0; i < BYTES_PER_PACKET; i++) {
        this->buffer[i] = buffer[offset + i];
    }

    windSampleCount = BitConverter::toInt16(this->buffer, NUM_WIND_SAMPLES_OFFSET);
    packetTime = extractArchiveDate();
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
const byte *
ArchivePacket::getBuffer() const {
    return buffer;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
int
ArchivePacket::getWindSampleCount() const {
    return windSampleCount;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
DateTime
ArchivePacket::getDateTime() const {
    return packetTime;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
bool ArchivePacket::isEmptyPacket() const {
    return packetTime == EMPTY_ARCHIVE_PACKET_TIME;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
DateTime
ArchivePacket::extractArchiveDate() const {
    int date = BitConverter::toInt16(buffer, DATE_STAMP_OFFSET);
    int time = BitConverter::toInt16(buffer, TIME_STAMP_OFFSET);
    int year = ((date >> 9) & 0x3F) + 2000;
    int month = (date >> 5) & 0xF;
    int day = date & 0x1F;
    int hour = time / 100;
    int minute = time % 100;

    time_t now = ::time(0);
    struct tm tm;
    Weather::localtime(now, tm);
    tm.tm_year = year - TIME_STRUCT_YEAR_OFFSET;
    tm.tm_mon = month - 1;
    tm.tm_mday = day;
    tm.tm_hour = hour;
    tm.tm_min = minute;
    tm.tm_sec = 0;
    return mktime(&tm);
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
std::string
ArchivePacket::formatMessage() const {
    ostringstream ss;
    ss << "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>";
    ss << "<historicalRecord>";
    DateTime archiveTime = extractArchiveDate();
    ss << "<time>" << Weather::formatDateTime(archiveTime) << "</time>";

    Measurement<Temperature> temperature;
    VP2Decoder::decode16BitTemperature(buffer, OUTSIDE_TEMPERATURE_OFFSET, temperature);
    ss << temperature.formatXML("avgOutdoorTemperature");

    VP2Decoder::decode16BitTemperature(buffer, HIGH_OUTSIDE_TEMPERATURE_OFFSET, temperature);
    ss << temperature.formatXML("highOutdoorTemperature");

    VP2Decoder::decode16BitTemperature(buffer, LOW_OUTSIDE_TEMPERATURE_OFFSET, temperature);
    ss << temperature.formatXML("lowOutdoorTemperature");

    Rainfall r = VP2Decoder::decodeRain(buffer, RAINFALL_OFFSET);
    ss << "<rainfall>" << r << "</rainfall>";

    r = VP2Decoder::decodeRain(buffer, HIGH_RAIN_RATE_OFFSET);
    ss << "<highRainfallRate>" << r << "</highRainfallRate>";

    Measurement<Pressure> baroPressure;
    VP2Decoder::decodeBarometricPressure(buffer, BAROMETER_OFFSET, baroPressure);
    ss << baroPressure.formatXML("baroPressure");

    Measurement<SolarRadiation> solarRadiation;
    VP2Decoder::decodeSolarRadiation(buffer, SOLAR_RADIATION_OFFSET, solarRadiation);
    ss << solarRadiation.formatXML("avgSolarRadiation");
  
    VP2Decoder::decode16BitTemperature(buffer, INSIDE_TEMPERATURE_OFFSET, temperature);
    ss << temperature.formatXML("indoorTemperature");

    Measurement<Humidity> humidity;
    VP2Decoder::decodeHumidity(buffer, INSIDE_HUMIDITY_OFFSET, humidity);
    ss << humidity.formatXML("indoorHumidity");

    VP2Decoder::decodeHumidity(buffer, OUTSIDE_HUMIDITY_OFFSET, humidity);
    ss << humidity.formatXML("outdoorHumidity");

    //
    // Both wind speed and direction must be valid to generate the XML
    //
    Measurement<Speed> windSpeed = VP2Decoder::decodeWindSpeed(buffer, AVG_WIND_SPEED_OFFSET);
    Measurement<Heading> windDir = VP2Decoder::decodeWindDirectionSlice(buffer, PREVAILING_WIND_DIRECTION_OFFSET);

    if (windSpeed.isValid() && windDir.isValid()) {
        ss << "<avgWind><speed>" << windSpeed << "</speed>"
           << "<direction>" << windDir << "</direction>"
           << "</avgWind>";
    }

    windSpeed = VP2Decoder::decodeWindSpeed(buffer, HIGH_WIND_SPEED_OFFSET);
    windDir = VP2Decoder::decodeWindDirectionSlice(buffer, DIR_OF_HIGH_WIND_SPEED_OFFSET);

    if (windSpeed.isValid() && windDir.isValid()) {
        ss << "<highWind><speed>" << windSpeed << "</speed>"
           << "<direction>" << windDir << "</direction>"
           << "</highWind>";
    }

    Measurement<UvIndex> uvIndex;
    VP2Decoder::decodeUvIndex(buffer, AVG_UV_INDEX_OFFSET, uvIndex);
    ss << uvIndex.formatXML("avgUvIndex");

    Measurement<Evapotranspiration> et = VP2Decoder::decodeDayET(buffer, ET_OFFSET);
    ss <<  et.formatXML("evapotranspiration");

    VP2Decoder::decodeSolarRadiation(buffer, HIGH_SOLAR_RADIATION_OFFSET, solarRadiation);
    ss << solarRadiation.formatXML("highSolarRadiation");

    VP2Decoder::decodeUvIndex(buffer, HIGH_UV_INDEX_OFFSET, uvIndex);
    ss << uvIndex.formatXML("highUvIndex");

    ss << "<extraHumidities>";
    for (int i = 0; i < MAX_EXTRA_HUMIDITIES; i++) {
        VP2Decoder::decodeHumidity(buffer, EXTRA_HUMIDITIES_BASE_OFFSET + i, humidity);
        if (humidity.isValid()) {
            ss << "<humidity><index>" << i << "</index><value>" << humidity.getValue() << "</value></humidity>";
        }
    }
    ss << "</extraHumidities>";

    ss << "<extraTemperatures>";
    for (int i = 0; i < MAX_EXTRA_TEMPERATURES; i++) {
        VP2Decoder::decode8BitTemperature(buffer, EXTRA_TEMPERATURES_BASE_OFFSET + i, temperature);
        if (temperature.isValid()) {
            ss << "<temperature><index>" << i << "</index><value>" << temperature.getValue() << "</value></temperature>";
        }
    }
    ss << "</extraTemperatures>";

/*
    for (int i = 0; i < VP2Constants::APB_MAX_LEAF_TEMPERATURES; i++) {
        value8 = buffer[LEAF_TEMPERATURE_BASE_OFFSET + i];
        if (value8 != VP2Constants::APB_INVALID_LEAF_TEMPERATURE)
            ss << ""; //leaf_temperature[{0}]={1};", i, value8 - EXTRA_TEMPERATURE_OFFSET);
    }

    ss << "<leafWetnessSensorEntries>";
    for (int i = 0; i < VP2Constants::APB_MAX_LEAF_WETNESSES; i++) {
        int leafWetness = BitConverter::toInt8(buffer, LEAF_WETNESS_BASE_OFFSET + i);
        if (leafWetness != VP2Constants::APB_INVALID_LEAF_WETNESS) {
            ss << "<entry><key>" << (500 + i) << "</key><value><sensorId>" << (500 + i) << "</sensorId><sensorType>LEAF_WETNESS</sensorType>";
            ss << "<measurement xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xsi:type=\"leafWetness\">";
            ss << leafWetness << "</measurement></value></entry>";
        }
    }
    ss << "</leafWetnessSensorEntries>";

    for (int i = 0; i < VP2Constants::APB_MAX_SOIL_TEMPERATURES; i++) {
        value8 = buffer[SOIL_TEMPERATURE_BASE_OFFSET + i];
        if (value8 != VP2Constants::APB_INVALID_SOIL_TEMPERATURE)
            ss << "";//soil_temperature[{0}]={1};", i, value8 - EXTRA_TEMPERATURE_OFFSET);
    }

    ss << "<soilMoistureSensorEntries>";
    for (int i = 0; i < VP2Constants::APB_MAX_SOIL_MOISTURES; i++) {
        int soilMoisture = BitConverter::toInt8(buffer, SOIL_MOISTURES_BASE_OFFSET + i);
        if (soilMoisture != VP2Constants::APB_INVALID_SOIL_MOISTURE) {
            ss << "<entry><key>" << (600 + i) << "</key><value><sensorId>" << (600 + i) << "</sensorId><sensorType>SOIL_MOISTURE</sensorType>";
            ss << "<measurement xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xsi:type=\"soilMoisture\">";
            ss << soilMoisture << "</measurement></value></entry>";
        }
    }
    ss << "</soilMoistureSensorEntries>";
*/
    ss << "</historicalRecord>";
    return ss.str();
}
}
